components/cloud/gcp/src/kms.rs

```diff
@@ -0,0 +1,407 @@
+// Copyright 2024 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::{fmt, result::Result as StdResult};
+
+use async_trait::async_trait;
+use cloud::{
+    error::{Error as CloudError, KmsError, Result},
+    kms::{Config, CryptographyType, DataKeyPair, EncryptedKey, KmsProvider, PlainKey},
+    metrics, KeyId,
+};
+use futures_util::stream::StreamExt;
+use http::Method;
+use hyper::Body;
+use regex::Regex;
+use serde::{Deserialize, Deserializer, Serialize};
+use tame_gcs::error::HttpStatusError;
+use tikv_util::{box_err, stream::RetryError, time::Instant};
+
+use crate::{
+    client::{GcpClient, RequestError},
+    STORAGE_VENDOR_NAME_GCP,
+};
+
+// generated random encryption data key length.
+const DEFAULT_DATAKEY_SIZE: usize = 32;
+// google kms endpoint.
+const GCP_KMS_ENDPOINT: &str = "https://cloudkms.googleapis.com/v1/";
+
+/// The encryption key_id pattern of gcp ksm:
+///   projects/{project_name}/locations/{location}/keyRings/{key_ring}/
+/// cryptoKeys/{key}
+const KEY_ID_PATTERN: &str =
+    r"^projects/([^/]+)/locations/([^/]+)/keyRings/([^/]+)/cryptoKeys/([^/]+)/?$";
```

Comment:
Is it possible to put a "/" or escaped "\/" in `project_name`?

Reply:
NO. From the [Resource name](https://google.aip.dev/122) google definition, it can't contains '/' char.